mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/master/resourcecluster/TaskExecutorAllocationHelper.java

```diff
@@ -0,0 +1,152 @@
+        this.allocationAttributesWithDefaults = Arrays.stream(allocationAttributesWithDefaults.split(","))
+            .filter(s -> !s.trim().isEmpty() && s.contains(":"))
+            .map(s -> s.split(":"))
+            .collect(Collectors.toMap(a -> a[0].trim(), a -> a[1].trim()));
+    }
+
+    public Optional<ContainerSkuID> findBestFitSkuID(AllocationConstraints allocationConstraints) {
+        ContainerSkuID bestFitSkuID = null;
+
+        for (SkuTypeSpec sku : skus.values()) {
+            if (canFit(sku.getSkuId(), allocationConstraints)) {
+                if (bestFitSkuID != null) {
+                    bestFitSkuID = findBestFit(bestFitSkuID, sku.getSkuId());
+                } else {
+                    bestFitSkuID = sku.getSkuId();
+                }
+            }
+        }
+        return bestFitSkuID == null ? Optional.empty() : Optional.of(bestFitSkuID);
+    }
+
+    public void addSkuIdIfAbsent(ContainerSkuID skuID) {
+        if (!skus.containsKey(skuID)) {
+            refreshMetadata(skuID);
+        }
+    }
+
+    public void removeSkuIdIfPresent(ContainerSkuID skuID) {
+        skus.remove(skuID);
+    }
+
+    private boolean canFit(ContainerSkuID skuID, AllocationConstraints constraints) {
+        return canFitResources(skuID, constraints) && areAllocationConstraintsSatisfied(skuID, constraints);
+    }
+
+    private boolean canFitResources(ContainerSkuID skuID, AllocationConstraints constraints) {
+        // TODO: shall we also check disk & other dims?
+        final SkuSizeSpec skuSize = skuSizes.get(getSkuSizeIdFrom(skus.get(skuID)));
+         return skuSize.getCpuCoreCount() >= constraints.getMachineDefinition().getCpuCores() &&
+             skuSize.getMemorySizeInMB() >= constraints.getMachineDefinition().getMemoryMB();
+    }
+
+    private boolean areAllocationConstraintsSatisfied(ContainerSkuID skuID, AllocationConstraints constraints) {
+        for (Entry<String, String> e : allocationAttributesWithDefaults.entrySet()) {
+            final String skuAttrValue = skus.get(skuID).getSkuMetadataFields().getOrDefault(e.getKey(), e.getValue());
+            final String constraintValue = constraints.getAssignmentAttributes().getOrDefault(e.getKey(), e.getValue());
+            if (!skuAttrValue.equalsIgnoreCase(constraintValue)) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    private ContainerSkuID findBestFit(ContainerSkuID sku0, ContainerSkuID sku1) {
+        final SkuSizeSpec size0 = skuSizes.get(getSkuSizeIdFrom(skus.get(sku0)));
+        final SkuSizeSpec size1 = skuSizes.get(getSkuSizeIdFrom(skus.get(sku1)));
+        if (size0.getCpuCoreCount() < size1.getCpuCoreCount() || (
+            size0.getCpuCoreCount() == size1.getCpuCoreCount() && size0.getMemorySizeInMB() < size1.getMemorySizeInMB()
+        )) {
+            return sku0;
+        }
+        return sku1;
+    }
+
+    private void refreshMetadata(ContainerSkuID skuID) {
```

Comment:
Hitting the database to refresh state is not recommended with actors, which are expected to manage the state. Can we manage this state either within the ResourceClusterActor or another actor that you think is more appropriate for this?

Reply:
agreed